src/v/cluster/plugin_frontend.cc

```diff
@@ -0,0 +1,378 @@
+ */
+#include "plugin_frontend.h"
+
+#include "cluster/controller_service.h"
+#include "cluster/controller_stm.h"
+
+namespace cluster {
+
+using model::transform_id;
+using model::transform_metadata;
+using model::transform_name;
+using mutation_result = plugin_frontend::mutation_result;
+namespace {
+errc map_errc(std::error_code ec) {
+    if (ec == errc::success) {
+        return errc::success;
+    }
+    if (ec.category() == raft::error_category()) {
+        switch (raft::errc(ec.value())) {
+        case raft::errc::timeout:
+            return errc::timeout;
+        case raft::errc::not_leader:
+            return errc::not_leader_controller;
+        default:
+            return errc::replication_error;
+        }
+    }
+    if (ec.category() == rpc::error_category()) {
+        switch (rpc::errc(ec.value())) {
+        case rpc::errc::client_request_timeout:
+            return errc::timeout;
+        default:
+            return errc::replication_error;
+        }
+    }
+    if (ec.category() == error_category()) {
+        return errc(ec.value());
+    }
+    return errc::replication_error;
+}
+} // namespace
+
+plugin_frontend::plugin_frontend(
+  model::node_id s,
+  partition_leaders_table* l,
+  plugin_table* t,
+  topic_table* tp,
+  controller_stm* c,
+  rpc::connection_cache* r,
+  ss::abort_source* a)
+  : _self(s)
+  , _leaders(l)
+  , _connections(r)
+  , _table(t)
+  , _topics(tp)
+  , _abort_source(a)
+  , _controller(c) {}
+
+ss::future<errc> plugin_frontend::upsert_transform(
+  transform_metadata meta, model::timeout_clock::time_point timeout) {
+    // The ID is looked up later.
+    transform_cmd c{transform_update_cmd{0, std::move(meta)}};
+    mutation_result r = co_await do_mutation(std::move(c), timeout);
+    co_return r.ec;
+}
+
+ss::future<mutation_result> plugin_frontend::remove_transform(
+  transform_name name, model::timeout_clock::time_point timeout) {
+    transform_cmd c{transform_remove_cmd{std::move(name), 0}};
+    return do_mutation(std::move(c), timeout);
+}
+
+ss::future<mutation_result> plugin_frontend::do_mutation(
+  transform_cmd cmd, model::timeout_clock::time_point timeout) {
+    auto cluster_leader = _leaders->get_leader(model::controller_ntp);
+    if (!cluster_leader) {
+        co_return mutation_result{.ec = errc::no_leader_controller};
+    }
+    if (*cluster_leader != _self) {
+        co_return co_await dispatch_mutation_to_remote(
+          *cluster_leader,
+          std::move(cmd),
+          timeout - model::timeout_clock::now());
+    }
+    if (ss::this_shard_id() != controller_stm_shard) {
+        co_return co_await container().invoke_on(
+          controller_stm_shard,
+          // NOLINTNEXTLINE(cppcoreguidelines-avoid-reference-coroutine-parameters)
+          [cmd = std::move(cmd), timeout](auto& service) mutable {
+              return service.do_mutation(std::move(cmd), timeout);
+          });
+    }
+    auto u = co_await _mu.get_units();
```

Comment:
nit: can be moved inside do_local_mutation()?

Reply:
Don't we need to hold the lock while we call insert_linearize_barrier?

Otherwise we could still be replicating (and I'm assuming applying the update in the plug-in backend) and then get stale results right?

Reply:
Or are you suggesting moving the barrier into do_local_mutation too? That SGTM and makes since because this function then is only about dispatching the command to the right place